src/basic/string-util.c

```diff
@@ -713,11 +714,13 @@ char *strip_tab_ansi(char **ibuf, size_t *_isz, size_t highlight[2]) {
                 }
         }
 
-        if (fflush_and_check(f) < 0) {
-                fclose(f);
-                return mfree(obuf);
-        }
-        fclose(f);
+        if (fflush_and_check(f) < 0)
+                return NULL;
+
+        f = safe_fclose(f);
+
+        if (!obuf)
+                return NULL;
```

Comment:
As stated in the commit message - fclose() does realloc() on the memstream and if that fails we're left with a NULL pointer, even if we do fflush() before:

```c
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>

void *realloc(void *ptr, size_t size) {
    return NULL;
}

int main(int argc, char *argv[])
{
    FILE *f;
    char *buf;
    size_t sz = 0;

    f = open_memstream(&buf, &sz);
    if (!f)
        return -ENOMEM;

    fflush(f);
    printf("buf: 0x%lx (%s), sz: %lu, errno: %d\n",
                (unsigned long) buf, buf, sz, errno);
    fclose(f);
    printf("buf: 0x%lx (%s), sz: %lu, errno: %d\n",
                (unsigned long) buf, buf, sz, errno);

    return 0;
}
```

```
$ gcc -o main main.c
$ ./main 
buf: 0x1f6a4a0 (), sz: 0, errno: 0
buf: 0x0 ((null)), sz: 0, errno: 0
```

I thought about adding a helper as well, so will prep a follow-up.

Reply:
Hmpf, but the helper won't *help* us much in this case, because even if the realloc() during fclose() fails, fclose() still returns 0, errno is also still 0, only the memstream pointer is NULL. This is pretty dumb...

Reply:
Nevermind, it *does* set the ERRNO, I'm just an idiot. Will prep a follow-up then.